### PR TITLE
Skip blocks without runnable tests

### LINKED ISSUE #2439: Block setup and teardown executed when tests are skipped
### Checklist

- [X] Issue has a meaningful title
- [X] I have searched the existing issues. See [all issues](https://github.com/pester/Pester/issues?q=is%3Aissue)
- [X] I have tested using the latest version of Pester. See [Installation and update guide](https://pester.dev/docs/introduction/installation).

### What is the issue?

`BeforeAll` and `AfterAll` are executed when all tests are skipped using `-Skip`.

This is a known limitation/backlog item as seen in this runtime test: https://github.com/pester/Pester/blob/7889a88acd0ecd632d725185a061eab78b4e4c9b/tst/Pester.Runtime.ts.ps1#L1228-L1275

### Expected Behavior

Setup and teardown are not executed when they're not needed

_Note: Tests skipped using `Set-ItResult` doesn't count as they're skipped during Run-phase (too late)_

### Steps To Reproduce

```powershell
$sb = {
    Describe 'A' {
        BeforeAll {
            Write-Host 'Before ALL -A'
        }
        BeforeEach {
            Write-Host 'Before EACH -A'
        }
        AfterEach {
            Write-Host 'After EACH -A'
        }

        It 'Test 1' -Skip {
            Write-Host 'This is test 1'
        }

        Describe 'B' {
            BeforeAll {
                Write-Host 'Before ALL -B'
            }
            AfterAll {
                Write-Host 'After ALL -B'
            }
            BeforeEach {
                Write-Host 'Before EACH -B'
            }
            AfterEach {
                Write-Host 'After EACH -B'
            }

            It 'Test 2' -Skip {
                Write-Host 'This is test 2'
            }
        }

        AfterAll {
            Write-Host 'After ALL -A'
        }
    }

}

$conf = New-PesterConfiguration
$conf.Output.Verbosity = 'Detailed'
$conf.Run.PassThru = $true
$conf.Run.ScriptBlock = $sb

$r = Invoke-Pester -Configuration $conf
```

Output:
```console
Pester v5.6.0-beta1

Starting discovery in 1 files.
Discovery found 2 tests in 35ms.
Running tests.
Before ALL -A
Describing A
  [!] Test 1 6ms (0ms|6ms)
Before ALL -B
 Describing B
   [!] Test 2 3ms (0ms|3ms)
After ALL -B
After ALL -A
Tests completed in 77ms
Tests Passed: 0, Failed: 0, Skipped: 2, Inconclusive: 0, NotRun: 0
```

### Describe your environment

Pester version     : 5.6.0-beta1 /workspaces/Pester/bin/Pester.psm1                                                     
PowerShell version : 7.4.1
OS version         : Unix 5.15.133.1

### Possible Solution?

_No response_

### LINKED ISSUE #2449: BeforeAll and AfterAll not invoked on skipped blocks when running test explicitly
### Checklist

- [X] Issue has a meaningful title
- [X] I have searched the existing issues. See [all issues](https://github.com/pester/Pester/issues?q=is%3Aissue)
- [X] I have tested using the latest version of Pester. See [Installation and update guide](https://pester.dev/docs/introduction/installation).

### What is the issue?

`BeforeAll` and `AfterAll` are not executed on skipped blocks when explicitly running a test, e.g. using `Filter.Line` option. See highlighted lines below which are missing.

### Expected Behavior

```diff
Discovery found 2 tests in 23ms.
Filter 'Line' set to ('C:\Users\Frode\Git\Pester/demoBlockSkip.tests.ps1:18').
Filters selected 1 tests to run.
Discovery: Test discovery finished.
Running tests.

Running tests from 'C:\Users\Frode\Git\Pester\demoBlockSkip.tests.ps1'
Before ALL ROOT
+Before ALL -A
+Before ALL -B
Describing A
 Describing B
Before EACH -A
Before EACH -B
This is test 2
After EACH -B
After EACH -A
   [+] Test 2 27ms (24ms|4ms)
+After ALL -B
+After ALL -A
After ALL ROOT
Tests completed in 101ms
Tests Passed: 1, Failed: 0, Skipped: 0, Inconclusive: 0, NotRun: 1
```

### Steps To Reproduce

demoBlockSkip.tests.ps1
```powershell
BeforeAll { Write-Host 'Before ALL ROOT' }
AfterAll { Write-Host 'After ALL ROOT' }

Describe 'A' -Skip {
    BeforeAll { Write-Host 'Before ALL -A' }
    AfterAll { Write-Host 'After ALL -A' }
    BeforeEach { Write-Host 'Before EACH -A' }
    AfterEach { Write-Host 'After EACH -A' }

    It 'Test 1' -Skip { Write-Host 'This is test 1' }

    Describe 'B' {
        BeforeAll { Write-Host 'Before ALL -B' }
        AfterAll { Write-Host 'After ALL -B' }
        BeforeEach { Write-Host 'Before EACH -B' }
        AfterEach { Write-Host 'After EACH -B' }

        It 'Test 2' -Skip { Write-Host 'This is test 2' }
    }
}
```

demoBlockSkip.ps1

```powershell
$conf = New-PesterConfiguration
$conf.Run.Path = "$PSScriptRoot/demoBlockSkip.tests.ps1"
$conf.Filter.Line = "$PSScriptRoot/demoBlockSkip.tests.ps1:18"
Invoke-Pester -Configuration $conf
```

### Describe your environment

Pester version     : 5.6.0-beta1 C:\Users\Frode\Documents\PowerShell\Modules\pester\5.6.0\Pester.psm1
PowerShell version : 7.4.2
OS version         : Microsoft Windows NT 10.0.22631.0

### Possible Solution?

Will be fixed by #2447 by setting `Skip` to false on blocks with child blocks/tests that will be executed. The logic needs to occur after evaluating all tests for explicit include recursively as tests usually inherit skip from parent block.

### PR DESCRIPTION
**Pull request by fflaten**
## PR Summary
Mark blocks (incl. container root) as Skip when no child tests will run. This is done to avoid running `BeforeAll/AfterAll` when it's not needed.

Fix #2439
Fix #2449

## PR Checklist

- [x] PR has meaningful title
- [x] Summary describes changes
- [x] PR is ready to be merged
  - If not, use the arrow next to `Create Pull Request` to mark it as a draft. PR can be marked `Ready for review` when it's ready.
- [x] Tests are added/update *(if required)*
- [ ] Documentation is updated/added *(if required)*

### CODE REVIEW

File: src/Pester.Runtime.ps1 [hunk 1]
```diff
@@ -2170,6 +2187,24 @@ function PostProcess-DiscoveredBlock {
...
+                        Write-PesterDebugMessage -Scope Skip "($($b.Path -join '.')) Block will be skipped because all children are excluded or skipped."
+                    }
+                }
+                $b.Skip = $true
+            } elseif ($b.Skip -and -not $shouldSkipBasedOnChildren) {
+                if ($PesterPreference.Debug.WriteDebugMessages.Value) {
+                    Write-PesterDebugMessage -Scope Skip "($($b.Path -join '.')) Block was marked as Skip, but one or more children should run and is not skipped, so the block will not be skipped."
+                }
+                $b.Skip = $false
+            }
```
Comment (fflaten) [inline src/Pester.Runtime.ps1:2206]: Why did I add this again? 😅

Comment (fflaten) [reply src/Pester.Runtime.ps1:2206]: Oh, it was to make BeforeAll/AfterAll run for blocks with `-Skip` if a child test will run due to Explicit include (line filter).

Pester 5.5.0

```powershell
Filter 'Line' set to ('C:\Users\Frode\Git\Pester/demoBlockSkip.tests.ps1:37').
Filters selected 1 tests to run.
Discovery: Test discovery finished.
Running tests.

Running tests from 'C:\Users\Frode\Git\Pester\demoBlockSkip.tests.ps1'
Before ALL ROOT
Describing A
 Describing B
Before EACH -A
Before EACH -B
This is test 2
After EACH -B
After EACH -A
   [+] Test 2 6ms (4ms|2ms)
After ALL ROOT
Tests completed in 72ms
Tests Passed: 1, Failed: 0, Skipped: 0, Inconclusive: 0, NotRun: 3
```

PR as-is

```powershell
Filter 'Line' set to ('/workspaces/Pester/demoBlockSkip.tests.ps1:37').
Filters selected 1 tests to run.
Discovery: Test discovery finished. 
Running tests.

Running tests from '/workspaces/Pester/demoBlockSkip.tests.ps1'
Before ALL ROOT
Before ALL -A
Before ALL -B
Describing A
 Describing B
Before EACH -A
Before EACH -B
This is test 2
After EACH -B
After EACH -A
   [+] Test 2 31ms (28ms|2ms)
After ALL -B
After ALL -A
After ALL ROOT
Tests completed in 145ms
Tests Passed: 1, Failed: 0, Skipped: 0 NotRun: 3
```

PR without the elseif-block (don't undo Skip):

```powershell
Filter 'Line' set to ('/workspaces/Pester/demoBlockSkip.tests.ps1:37').
Filters selected 1 tests to run.
Discovery: Test discovery finished. 
Running tests.

Running tests from '/workspaces/Pester/demoBlockSkip.tests.ps1'
Describing A
 Describing B
Before EACH -A
Before EACH -B
This is test 2
After EACH -B
After EACH -A
   [+] Test 2 53ms (25ms|28ms)
Tests completed in 413ms
Tests Passed: 1, Failed: 0, Skipped: 0 NotRun: 3
```

Which scenario is correct?

Comment (nohwnd) [reply src/Pester.Runtime.ps1:2206]: Yes, that was the idea. Test is broken you skip it, but you can still run it if you run just that. Another alternative is using a tag, but then you have to always exclude that tag.

Comment (fflaten) [reply src/Pester.Runtime.ps1:2206]: Cool. I'll keep the fix to run all before-/afterall blocks 👍

We need to get this logic documented before v6.

**Comment by fflaten** [comment]
Ready for review, but merge pending updated base branch for v6

**Comment by fflaten** [comment]
@nohwnd Target this for 5.6.0 instead or wait? I'm fine with both, but shouldn't be a breaking change.

**Review by nohwnd — APPROVED** [review]

**Pull request merged**